Mathlib/Combinatorics/SimpleGraph/Metric.lean

```diff
@@ -43,105 +41,190 @@ variable {V : Type*} (G : SimpleGraph V)
 
 /-! ## Metric -/
 
+section edist
+
+/--
+The extended distance between two vertices is the length of the shortest walk between them.
+It is `⊤` if no such walk exists.
+-/
+noncomputable def edist (u v : V) : ℕ∞ :=
+  ⨅ w : G.Walk u v, w.length
+
+variable {G} {u v w : V}
+
+protected theorem Reachable.exists_walk_of_edist (hr : G.Reachable u v) :
+    ∃ p : G.Walk u v, p.length = G.edist u v :=
+  csInf_mem <| Set.range_nonempty_iff_nonempty.mpr hr
+
+protected theorem Connected.exists_walk_of_edist (hconn : G.Connected) (u v : V) :
+    ∃ p : G.Walk u v, p.length = G.edist u v :=
+  (hconn u v).exists_walk_of_edist
+
+theorem edist_le (p : G.Walk u v) :
+    G.edist u v ≤ p.length :=
+  sInf_le ⟨p, rfl⟩
+
+@[simp]
+theorem edist_eq_zero_iff_eq :
+    G.edist u v = 0 ↔ u = v := by
+  apply Iff.intro <;> simp [edist, ENat.iInf_eq_zero]
+
+theorem edist_self : edist G v v = 0 :=
+  edist_eq_zero_iff_eq.mpr rfl
+
+theorem pos_edist_of_ne (hne : u ≠ v) :
+    0 < G.edist u v :=
+  pos_iff_ne_zero.mpr <| edist_eq_zero_iff_eq.ne.mpr hne
+
+lemma edist_eq_top_of_not_reachable (h : ¬G.Reachable u v) :
+    G.edist u v = ⊤ := by
+  simp [edist, not_reachable_iff_isEmpty_walk.mp h]
+
+theorem Reachable.of_edist_ne_top (h : G.edist u v ≠ ⊤) :
+    G.Reachable u v :=
+  not_not.mp <| edist_eq_top_of_not_reachable.mt h
+
+protected theorem Connected.edist_triangle (hconn : G.Connected) :
+    G.edist u w ≤ G.edist u v + G.edist v w := by
+  obtain ⟨p, hp⟩ := hconn.exists_walk_of_edist u v
+  obtain ⟨q, hq⟩ := hconn.exists_walk_of_edist v w
+  rw [← hp, ← hq, ← Nat.cast_add, ← Walk.length_append]
+  apply edist_le
+
+theorem edist_comm : G.edist u v = G.edist v u := by
+  have {u v : V} (h : G.Reachable u v) : G.edist u v ≤ G.edist v u := by
+    obtain ⟨p, hp⟩ := h.symm.exists_walk_of_edist
+    rw [← hp, ← Walk.length_reverse]
+    apply edist_le
+  by_cases h : G.Reachable u v
+  · apply le_antisymm (this h) (this h.symm)
+  · have h' : ¬G.Reachable v u := fun h' => absurd h'.symm h
+    simp [h, h', edist_eq_top_of_not_reachable]
+
+lemma exists_walk_of_edist_ne_top (h : G.edist u v ≠ ⊤) :
+    ∃ p : G.Walk u v, p.length = G.edist u v :=
+  (Reachable.of_edist_ne_top h).exists_walk_of_edist
+
+/--
+The extended distance between vertices is equal to `1` if and only if these vertices are adjacent.
+-/
+theorem edist_eq_one_iff_adj : G.edist u v = 1 ↔ G.Adj u v := by
+  have ne_top_of_eq_one {n : ℕ∞} (h : n = 1) : n ≠ ⊤ := by
+    rw [← ENat.coe_toNat_eq_self, h]; rfl
+  refine ⟨fun h ↦ ?_, fun h ↦ ?_⟩
+  · let ⟨w, hw⟩ := exists_walk_of_edist_ne_top <| ne_top_of_eq_one h
+    rw [h, Nat.cast_eq_one] at hw
+    exact w.adj_of_length_eq_one <| hw
+  · exact ge_antisymm (ENat.one_le_iff_pos.mpr <| pos_edist_of_ne h.ne) <| edist_le h.toWalk
```

Comment:
Changes:
* Use `obtain` instead of `let` since it's a proof (`have` would work too). There's no difference in this case, but `let` signals "I care about the definition"
* Rather than define a sub-lemma, it's possible to use `simp` (or `decide`) to handle the neq. This is a slight abuse of `rw`'s ability to unify metavariables, and more principled would be `obtain ⟨w, hw⟩ := exists_walk_of_edist_ne_top (by rw [h]; simp : G.edist u v ≠ ⊤)`.
* `le_antisymm` instead of `ge_antisymm` since the operands are in terms of le, for clarity.

```suggestion
theorem edist_eq_one_iff_adj : G.edist u v = 1 ↔ G.Adj u v := by
  refine ⟨fun h ↦ ?_, fun h ↦ ?_⟩
  · obtain ⟨w, hw⟩ := exists_walk_of_edist_ne_top <| by rw [h]; simp
    rw [h, Nat.cast_eq_one] at hw
    exact w.adj_of_length_eq_one hw
  · exact le_antisymm (edist_le h.toWalk) (ENat.one_le_iff_pos.mpr <| pos_edist_of_ne h.ne)
```